File: app/src/main/java/com/weiqi/ui/board/MiniStone.kt
package com.weiqi.ui.board

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.weiqi.engine.StoneColor
import com.weiqi.ui.theme.Zen

/** Small standalone stone, e.g. for player avatars in lists. */
@Composable
fun MiniStone(color: StoneColor, modifier: Modifier = Modifier, size: Dp = 24.dp) {
    Canvas(modifier = modifier.size(size)) {
        val r = this.size.minDimension / 2f
        val center = Offset(this.size.width / 2f, this.size.height / 2f)
        val isBlack = color == StoneColor.BLACK
        val top = if (isBlack) Zen.blackStoneTop else Zen.whiteStoneTop
        val bot = if (isBlack) Zen.blackStoneBottom else Zen.whiteStoneBottom
        // Ambient shadow.
        drawCircle(Color.Black.copy(alpha = 0.20f), radius = r,
            center = center.copy(y = center.y + r * 0.18f))
        drawCircle(
            brush = Brush.radialGradient(
                0f to top, 1f to bot,
                center = Offset(center.x - r * 0.3f, center.y - r * 0.3f),
                radius = r * 1.4f
            ),
            radius = r, center = center
        )
        drawCircle(
            color = if (isBlack) Color.Black else Zen.gridInk.copy(alpha = 0.55f),
            radius = r, center = center,
            style = Stroke(width = if (isBlack) 0.6f else 1f)
        )
    }
}

File: lib/src/ui/board/mini_stone.dart
import 'package:flutter/material.dart';

import '../../domain/models.dart';
import '../theme.dart';

/// Small standalone stone, e.g. for player avatars in lists.
class MiniStone extends StatelessWidget {
  final StoneColor color;
  final double size;

  const MiniStone({super.key, required this.color, this.size = 24});

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: size,
      height: size,
      child: CustomPaint(painter: _MiniStonePainter(color)),
    );
  }
}

class _MiniStonePainter extends CustomPainter {
  final StoneColor color;
  _MiniStonePainter(this.color);

  @override
  void paint(Canvas canvas, Size size) {
    final r = size.shortestSide / 2;
    final center = Offset(size.width / 2, size.height / 2);
    final isBlack = color == StoneColor.black;
    final top = isBlack ? Zen.blackStoneTop : Zen.whiteStoneTop;
    final bot = isBlack ? Zen.blackStoneBottom : Zen.whiteStoneBottom;

    canvas.drawCircle(
      center.translate(0, r * 0.18),
      r,
      Paint()..color = const Color(0x33000000),
    );

    final gradient = RadialGradient(
      center: const Alignment(-0.3, -0.3),
      radius: 0.7,
      colors: [top, bot],
    );
    final rect = Rect.fromCircle(center: center, radius: r);
    canvas.drawCircle(
      center,
      r,
      Paint()..shader = gradient.createShader(rect),
    );

    canvas.drawCircle(
      center,
      r,
      Paint()
        ..style = PaintingStyle.stroke
        ..color = isBlack
            ? Colors.black.withValues(alpha: 0.80)
            : Zen.gridInk.withValues(alpha: 0.42)
        ..strokeWidth = isBlack ? 0.45 : 0.7,
    );
  }

  @override
  bool shouldRepaint(covariant _MiniStonePainter oldDelegate) =>
      oldDelegate.color != color;
}
